api/src/main/java/org/pmiops/workbench/billing/FreeTierBillingService.java

```diff
@@ -25,55 +35,229 @@
       WorkspaceFreeTierUsageDao workspaceFreeTierUsageDao,
-      Provider<WorkbenchConfig> workbenchConfigProvider) {
+      Provider<WorkbenchConfig> workbenchConfigProvider,
+      Clock clock) {
     this.bigQueryService = bigQueryService;
     this.notificationService = notificationService;
+    this.userDao = userDao;
     this.workspaceDao = workspaceDao;
     this.workspaceFreeTierUsageDao = workspaceFreeTierUsageDao;
     this.workbenchConfigProvider = workbenchConfigProvider;
+    this.clock = clock;
   }
 
+  /**
+   * Check whether users have incurred sufficient cost or time in their workspaces to trigger alerts
+   * due to passing thresholds or exceeding limits
+   */
   public void checkFreeTierBillingUsage() {
-    final Map<DbWorkspace, Double> workspaceCosts = getFreeTierWorkspaceCosts();
+    // retrieve the costs stored in the DB from the last time this was run
+    final Map<DbUser, Double> previousUserCosts = workspaceFreeTierUsageDao.getUserCostMap();
 
-    final Set<DbUser> expiredCreditsUsers =
+    // retrieve current workspace costs from BigQuery and store in the DB
+    final Map<DbWorkspace, Double> workspaceCosts = getFreeTierWorkspaceCostsFromBQ();
+    workspaceCosts.forEach(workspaceFreeTierUsageDao::updateCost);
+
+    // sum current workspace costs by workspace creator
+    final Map<DbUser, Double> userCosts =
         workspaceCosts.entrySet().stream()
             .collect(
                 Collectors.groupingBy(
-                    e -> e.getKey().getCreator(), Collectors.summingDouble(Entry::getValue)))
-            .entrySet()
-            .stream()
-            .filter(entry -> entry.getValue() > getUserFreeTierDollarLimit(entry.getKey()))
+                    e -> e.getKey().getCreator(), Collectors.summingDouble(Entry::getValue)));
+
+    // check cost and time thresholds for the relevant users
+
+    // collect previously-expired and currently-expired users by cost and time
+    // for users which are expired: alert only if they were not expired previously
+    // for users which are not yet expired:
+    //    check for intermediate thresholds and alert, possibly for both cost and time
+
+    final Set<DbUser> previouslyExpiredUsers = getExpiredUsersFromDb();
+
+    final Set<DbUser> costExpiredUsers =
+        userCosts.entrySet().stream()
+            .filter(this::expiredByCost)
             .map(Entry::getKey)
             .collect(Collectors.toSet());
 
-    for (DbUser expiredUser : expiredCreditsUsers) {
-      notificationService.alertUser(expiredUser, "You have exceeded your free tier credits.");
+    final Set<DbUser> timeExpiredUsers =
+        userDao.findByFirstRegistrationCompletionTimeNotNull().stream()
+            .filter(this::expiredByTime)
+            .collect(Collectors.toSet());
+
+    final Set<DbUser> currentExpiredUsers = Sets.union(costExpiredUsers, timeExpiredUsers);
+    processNewlyExpiredUsers(Sets.difference(currentExpiredUsers, previouslyExpiredUsers));
```

Comment:
nit: I feel like this function name doesn't tell me much since I already know that the input argument is the set of newly expired users, `Sets.difference(currentExpiredUsers, previouslyExpiredUsers)`. Additionally, I think function names can be confusing if they're named after how they should used instead of what they are doing. In this example, processing newly expired users is only correct if the caller passes that in. The function itself doesn't check anything regarding the newly expired status.

What do you think of something like this?
```
Set<DbUser> newlyExpiredUsers = Sets.difference(currentExpiredUsers, previouslyExpiredUsers);
deactiveUserWorkspaces(newlyExpiredUsers);

  private void deactivateUserWorkspaces(Set<DbUser> users) {
    for (final DbUser user : users) {
      notificationService.alertUserFreeTierExpiration(user);
      final Set<DbWorkspace> toDeactivate = workspaceDao.findAllByCreator(user);
      for (final DbWorkspace workspace : toDeactivate) {
        workspaceDao.updateBillingStatus(workspace.getWorkspaceId(), BillingStatus.INACTIVE);
      }
    }
  }
```

or

```
Set<DbUser> newlyExpiredUsers = Sets.difference(currentExpiredUsers, previouslyExpiredUsers);
newlyExpiredUsers.forEach(user -> {
  notificationService.alertUserFreeTierExpiration(user);
  deactiveUserWorkspaces(user);
});

private void deactivateUserWorkspaces(DbUser user) {
    final Set<DbWorkspace> toDeactivate = workspaceDao.findAllByCreator(user);
    for (final DbWorkspace workspace : toDeactivate) {
      workspaceDao.updateBillingStatus(workspace.getWorkspaceId(), BillingStatus.INACTIVE);
    }
  }
```

Reply:
I like that, thanks.